# 1.只有同步功能

1.每个 promise 都有三个状态 pennding 等待态 resolve 标识变成成功态 fulfilled reject 标识变成失败态 REJECTED

2.每个 promise 需要有一个 then 方法 ， 传入两个参数 一个是成功的回调另一个是失败的回调

3.new Promise 会立即执行

4.一旦成功就不能失败 一旦失败不能成功

5.当 promise 抛出异常后 也会走失败态

```js
const enum STATUS { // 存放状态，ts枚举
  pending = 'PENDING',
  fulfilled = 'FULFILLED',
  rejected = 'REJECTED'
}

class Promise {
  status: STATUS
  value: any
  reason: any
  constructor(executor) {
    this.status = STATUS.pending; // 当前状态
    this.value = undefined; // 成功原因
    this.reason = undefined;  // 失败原因

    const resolve = (value?: any) => { // 调resolve更改status状态，并把传入的value保存下来。加？表示value允许不传参
      if (this.status === STATUS.pending) { // 只有等待态时才允许修改状态和值，否则不生效
        this.status = STATUS.fulfilled
        this.value = value
      }
    }
    const reject = (reason?: any) => { // 逻辑和resolve类似
      if (this.status === STATUS.pending) {
        this.status = STATUS.rejected
        this.reason = reason
      }
    }

    try { // 如果promise初始执行报错，也走reject状态
      executor(resolve, reject)
    } catch (e) {
      reject(e)
    }
  }
  then(onFulfilled, onRejected) {
    if (this.status == STATUS.fulfilled) {
      onFulfilled(this.value)
    }
    if (this.status == STATUS.rejected) {
      onRejected(this.reason)
    }
  }

}


export default Promise
```

测试代码

```js
const Promise = require('./bundle');

let promise = new Promise((resolve, reject) => {
  resolve('ok');
  // reject('err1')
  // throw new Error('err2')
});

promise.then(
  data => {
    console.log('success', data);
  },
  err => {
    console.log('faild', err);
  },
);

// success ok
```

总结：

实现了 promise 的状态转换和传参

缺陷是只支持同步功能，new Promise 里有异步代码（定时器）就失效了

解决方法是如果 STATUS 是等待态，调用 then 方法时需要把成功回调和失败回调存起来，resolve()、reject()方法时遍历成功回调和失败回调。这是发布订阅模式在 promise 里的应用。
